File: leptos-kanban/src/types/modals.rs
use crate::types::{serialize::ColumnType, state::{BoardState, ColumnState, TaskState}};


#[derive(Clone)]
pub struct OpenTaskModal {
    pub column_id: Option<ColumnType>,
    pub task: Option<TaskState>,
}

impl OpenTaskModal {
    pub fn new(column_id: Option<ColumnType>) -> Self {
        Self { column_id, task: None }
    }

    pub fn new_with_task(column_id: ColumnType, task: TaskState) -> Self {
        Self { column_id: Some(column_id), task: Some(task) }
    }
}

#[derive(Clone)]
pub struct OpenColumnModal {
    pub column: ColumnState,
}

impl OpenColumnModal {
    pub fn new_with_column(column: ColumnState) -> Self {
        Self { column }
    }
}

#[derive(Clone)]
pub struct OpenBoardModal {
    pub board: Option<BoardState>,
}

impl OpenBoardModal {
    pub fn new() -> Self {
        Self { board: None }
    }

    pub fn new_with_board(board: BoardState) -> Self {
        Self { board: Some(board) }
    }
}

File: react-kanban/src/types/modals.ts
import type { Board, Column, ColumnType, Task } from "./serialize";

export type OpenTaskModal = {
  columnType: ColumnType | null;
  task: Task | null;
};

export function openTaskModal(columnType: ColumnType | null): OpenTaskModal {
  return {
    columnType,
    task: null,
  };
}

export function openTaskModalWithTask(
  columnType: ColumnType,
  task: Task,
): OpenTaskModal {
  return {
    columnType,
    task,
  };
}

export type OpenColumnModal = {
  column: Column;
};

export function openColumnModalWithColumn(column: Column): OpenColumnModal {
  return { column };
}

export type OpenBoardModal = {
  board: Board | null;
};

export function openBoardModal(): OpenBoardModal {
  return { board: null };
}

export function openBoardModalWithBoard(board: Board): OpenBoardModal {
  return { board };
}
